src/frequenz/client/reporting/_client.py

```diff
@@ -0,0 +1,289 @@
+
+        page_token = None
+
+        while True:
+            pagination_params = PBPaginationParams(
+                page_size=page_size, page_token=page_token
+            )
+
+            response = await self._fetch_page(
+                microgrid_components=microgrid_components_pb,
+                metrics=metrics_pb,
+                list_filter=list_filter,
+                pagination_params=pagination_params,
+            )
+            if not response or response.is_empty():
+                break
+
+            yield response
+
+            page_token = response.next_page_token
+            if not page_token:
+                break
+
+    async def _fetch_page(
+        self,
+        *,
+        microgrid_components: list[PBMicrogridComponentIDs],
+        metrics: list[PBMetric.ValueType],
+        list_filter: PBListMicrogridComponentsDataRequest.ListFilter,
+        pagination_params: PBPaginationParams,
+    ) -> ComponentsDataPage | None:
+        """Fetch a single page of microgrid components data.
+
+        Args:
+            microgrid_components: A list of microgrid components.
+            metrics: A list of metrics.
+            list_filter: A list filter.
+            pagination_params: A pagination params.
+
+        Returns:
+            A ComponentsDataPage object of microgrid components data.
+        """
+        try:
+            request = PBListMicrogridComponentsDataRequest(
+                microgrid_components=microgrid_components,
+                metrics=metrics,
+                filter=list_filter,
+                pagination_params=pagination_params,
+            )
+            response = await cast(
+                Awaitable[PBListMicrogridComponentsDataResponse],
+                self._stub.ListMicrogridComponentsData(request),
+            )
+        except grpcaio.AioRpcError as e:
+            print(f"RPC failed: {e}")
+            return None
+        return ComponentsDataPage(response)
+
+    async def close(self) -> None:
+        """Close the client and cancel any pending requests immediately."""
+        await self._grpc_channel.close(grace=None)
+
+    async def __aenter__(self) -> "ReportingClient":
+        """Enter the async context."""
+        return self
+
+    async def __aexit__(
+        self,
+        _exc_type: Type[BaseException] | None,
+        _exc_val: BaseException | None,
+        _exc_tb: Any | None,
+    ) -> bool | None:
+        """
+        Exit the asynchronous context manager.
+
+        Note that exceptions are not handled here, but are allowed to propagate.
+
+        Args:
+            _exc_type: Type of exception raised in the async context.
+            _exc_val: Exception instance raised.
+            _exc_tb: Traceback object at the point where the exception occurred.
+
+        Returns:
+            None, allowing any exceptions to propagate.
+        """
+        await self.close()
+        return None
```

Comment:
This might be a good foundation for a `BaseApiClient` class that can live in https://github.com/frequenz-floss/frequenz-client-base-python/!

It could also standarize the constructor, I think right now some clients take a grpc channel and some take a connection string. I think we should go with the later to remove the explicit dependency on grpc for downstream projects.

Reply:
* https://github.com/frequenz-floss/frequenz-client-base-python/issues/42

Reply:
Sounds good

Reply:
https://github.com/frequenz-floss/frequenz-client-reporting-python/issues/30